### ISSUE TITLE
大模型相关调研

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#9557]
大模型相关调研<br>
<br>
https://ift.tt/EFXSc78<br>
<br>
Ph.D. Candidate at HUST<br>
<br>

<div><div><div><div><div><pre><code><span>def</span> <span>main</span><span>():</span>
  
    <span># Disable CuDNN.
</span>    <span>torch</span><span>.</span><span>backends</span><span>.</span><span>cudnn</span><span>.</span><span>enabled</span> <span>=</span> <span>False</span>
  
    <span># Timer.
</span>    <span>timers</span> <span>=</span> <span>Timers</span><span>()</span>
  
    <span># Arguments.
</span>    <span>args</span> <span>=</span> <span>get_args</span><span>()</span>
  
    <span>os</span><span>.</span><span>makedirs</span><span>(</span><span>args</span><span>.</span><span>save</span><span>,</span> <span>exist_ok</span><span>=</span><span>True</span><span>)</span>
  
    <span># Pytorch distributed.
</span>    <span>initialize_distributed</span><span>(</span><span>args</span><span>)</span>
    <span>if</span> <span>torch</span><span>.</span><span>distributed</span><span>.</span><span>get_rank</span><span>()</span> <span>==</span> <span>0</span><span>:</span>
        <span>print</span><span>(</span><span>'Pretrain Enc-Dec model'</span><span>)</span>
        <span>print_args</span><span>(</span><span>args</span><span>)</span>
        <span>with</span> <span>open</span><span>(</span><span>os</span><span>.</span><span>path</span><span>.</span><span>join</span><span>(</span><span>args</span><span>.</span><span>save</span><span>,</span> <span>"args.json"</span><span>),</span> <span>"w"</span><span>)</span> <span>as</span> <span>f</span><span>:</span>
            <span>json</span><span>.</span><span>dump</span><span>(</span><span>vars</span><span>(</span><span>args</span><span>),</span> <span>f</span><span>)</span>
  
    <span># Random seeds for reproducability.
</span>    <span>set_random_seed</span><span>(</span><span>args</span><span>.</span><span>seed</span><span>)</span>
  
    <span># setup tokenizer
</span>    <span>tokenizer</span> <span>=</span> <span>EncDecTokenizer</span><span>(</span><span>os</span><span>.</span><span>path</span><span>.</span><span>join</span><span>(</span><span>args</span><span>.</span><span>tokenizer_path</span><span>,</span> <span>'vocab.txt'</span><span>))</span>
      
    <span>with</span> <span>open</span><span>(</span><span>args</span><span>.</span><span>deepspeed_config</span><span>,</span> <span>"r"</span><span>)</span> <span>as</span> <span>f</span><span>:</span>
        <span>ds_config</span> <span>=</span> <span>json</span><span>.</span><span>load</span><span>(</span><span>f</span><span>)</span>
  
    <span>args</span><span>.</span><span>gradient_accumulation_steps</span> <span>=</span> <span>ds_config</span><span>[</span><span>"gradient_accumulation_steps"</span><span>]</span>
    <span>args</span><span>.</span><span>zero_stage</span> <span>=</span> <span>ds_config</span><span>[</span><span>'zero_optimization'</span><span>][</span><span>'stage'</span><span>]</span>
  
    <span># Model, optimizer, and learning rate.
</span>    <span>model</span><span>,</span> <span>optimizer</span><span>,</span> <span>lr_scheduler</span> <span>=</span> <span>setup_model_and_optimizer</span><span>(</span><span>args</span><span>,</span> <span>tokenizer</span><span>.</span><span>vocab_size</span><span>)</span>
    <span>optimizer</span><span>.</span><span>cur_scale</span> <span>=</span> <span>4096</span>
      
    <span>if</span> <span>torch</span><span>.</span><span>distributed</span><span>.</span><span>get_rank</span><span>()</span> <span>==</span> <span>0</span><span>:</span>
        <span>print</span><span>(</span><span>args</span><span>.</span><span>iteration</span><span>)</span>
      
    <span># 数据集读取和划分
</span>    <span>train_data_iterator</span><span>,</span> <span>val_data_iterator</span><span>,</span> <span>test_data_iterator</span> <span>=</span> \
            <span>build_train_valid_test_data_iterators</span><span>(</span>
                    <span>train_valid_test_dataset_provider</span><span>,</span> <span>args</span><span>,</span> <span>tokenizer</span><span>)</span>
  
    <span>#训练
</span>    <span>iteration</span> <span>=</span> <span>0</span>
    <span>if</span> <span>args</span><span>.</span><span>train_iters</span> <span>&gt;</span> <span>0</span><span>:</span>
        <span>iteration</span><span>,</span> <span>skipped</span> <span>=</span> <span>train</span><span>(</span><span>tokenizer</span><span>,</span> <span>model</span><span>,</span> <span>optimizer</span><span>,</span>
                                   <span>lr_scheduler</span><span>,</span>
                                   <span>train_data_iterator</span><span>,</span>
                                   <span>val_data_iterator</span><span>,</span>
                                   <span>timers</span><span>,</span> <span>args</span><span>)</span>
  
        <span>prefix</span> <span>=</span> <span>'the end of training for val data'</span>
        <span>evaluate_and_print_results</span><span>(</span><span>tokenizer</span><span>,</span> <span>prefix</span><span>,</span> <span>val_data_iterator</span><span>,</span>
                                                  <span>model</span><span>,</span> <span>args</span><span>,</span> <span>timers</span><span>,</span> <span>False</span><span>)</span>
  	
    <span>#模型保存 checkpoint
</span>    <span>if</span> <span>args</span><span>.</span><span>save</span> <span>and</span> <span>iteration</span> <span>!=</span> <span>0</span><span>:</span>
        <span>save_checkpoint</span><span>(</span><span>iteration</span><span>,</span> <span>model</span><span>,</span> <span>optimizer</span><span>,</span> <span>lr_scheduler</span><span>,</span> <span>args</span><span>)</span>
  
    <span>if</span> <span>args</span><span>.</span><span>do_test</span><span>:</span>
        <span># Run on test data.
</span>        <span>prefix</span> <span>=</span> <span>'the end of training for test data'</span>
        <span>evaluate_and_print_results</span><span>(</span><span>tokenizer</span><span>,</span> <span>prefix</span><span>,</span> <span>test_data_iterator</span><span>,</span>
                                   <span>model</span><span>,</span> <span>args</span><span>,</span> <span>timers</span><span>,</span> <span>True</span><span>)</span>
</code></pre></div></div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via Emperorlu's Site <br>
<br>
June 25, 2023 at 11:35AM
